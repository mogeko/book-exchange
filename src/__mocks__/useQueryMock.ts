import * as hooks from "@/lib/hooks/useQuery";

const mock = jest.spyOn(hooks, "default");

const useQueryMock = {
  target: hooks.default,
  returnResult: (res = {}) => mock.mockImplementation(() => genExampleRes(res)),
};

export function genExampleRes(res: Partial<ResType> = {}) {
  const exampleRes = {
    data: undefined,
    isValidating: false,
    isError: false,
    isLoading: false,
    mutate: jest.fn(),
  };

  return { ...exampleRes, ...res };
}

type ResType = hooks.QueryReturn;

export default useQueryMock;
