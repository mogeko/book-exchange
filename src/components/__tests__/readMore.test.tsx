import { render, screen, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";
import { faker } from "@faker-js/faker";
import useQueryMock from "@/__mocks__/useQueryMock";
import withReadMore from "@/components/readMore";

const ReadMore = withReadMore(({ children }) => <div>{children}</div>);
const exampleData = faker.lorem.paragraph(50);
const exampleFoldedData = {
  text: faker.lorem.paragraph(10),
  is_folded: true,
};

describe("ReadMore", () => {
  beforeEach(() => {
    useQueryMock.returnResult({ data: { text: exampleData } });
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it("renders a ReadMore without ReadMore button", () => {
    const { container } = render(
      <ReadMore
        foldedData={{ ...exampleFoldedData, is_folded: false }}
        url="/example/path"
      />
    );

    expect(screen.getByText(exampleFoldedData.text)).toBeInTheDocument();
    expect(container.querySelector("button")).not.toBeInTheDocument();
  });

  it("snapshots a ReadMore without ReadMore button", () => {
    const { container } = render(
      <ReadMore
        foldedData={{ ...exampleFoldedData, is_folded: false }}
        url="/example/path"
      />
    );

    expect(container).toMatchSnapshot();
  });

  it("renders a ReadMore with ReadMore button", () => {
    const { container } = render(
      <ReadMore foldedData={exampleFoldedData} url="/example/path" />
    );

    expect(screen.getByText(exampleFoldedData.text)).toBeInTheDocument();
    expect(container.querySelector("button")?.textContent).toBe(" Read more");
  });

  it("snapshots a ReadMore with ReadMore button", () => {
    const { container } = render(
      <ReadMore foldedData={exampleFoldedData} url="/example/path" />
    );

    expect(container).toMatchSnapshot();
  });

  it("renders a ReadMore with ReadLess button", () => {
    const { container } = render(
      <ReadMore foldedData={exampleFoldedData} url="/example/path" />
    );
    fireEvent.click(container.querySelector("button")!);

    expect(screen.getByText(exampleData)).toBeInTheDocument();
    expect(container.querySelector("button")?.textContent).toBe(" Read less");
  });

  it("snapshots a ReadMore with ReadLess button", () => {
    const { container } = render(
      <ReadMore foldedData={exampleFoldedData} url="/example/path" />
    );
    fireEvent.click(container.querySelector("button")!);

    expect(container).toMatchSnapshot();
  });
});

describe("ReadMore with abnormal state", () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it("renders a ReadMore when loading", () => {
    useQueryMock.returnResult({ isLoading: true });
    const { container } = render(
      <ReadMore foldedData={exampleFoldedData} url="/example/path" />
    );
    fireEvent.click(container.querySelector("button")!);

    expect(screen.getByText(exampleFoldedData.text)).toBeInTheDocument();
    expect(container.querySelector("button")?.textContent).toBe("Loading...");
  });

  it("snapshots a ReadMore when loading", () => {
    useQueryMock.returnResult({ isLoading: true });
    const { container } = render(
      <ReadMore foldedData={exampleFoldedData} url="/example/path" />
    );
    fireEvent.click(container.querySelector("button")!);

    expect(container).toMatchSnapshot();
  });

  it("renders a ReadMore when error", () => {
    useQueryMock.returnResult({ isError: true });
    const { container } = render(
      <ReadMore foldedData={exampleFoldedData} url="/example/path" />
    );
    fireEvent.click(container.querySelector("button")!);

    expect(screen.getByText("Network Error!")).toBeInTheDocument();
    expect(container.querySelector("button")).not.toBeInTheDocument();
  });

  it("snapshots a ReadMore when error", () => {
    useQueryMock.returnResult({ isError: true });
    const { container } = render(
      <ReadMore foldedData={exampleFoldedData} url="/example/path" />
    );
    fireEvent.click(container.querySelector("button")!);

    expect(container).toMatchSnapshot();
  });
});
