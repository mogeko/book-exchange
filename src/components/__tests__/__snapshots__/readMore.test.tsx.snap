// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReadMore snapshots a ReadMore with ReadLess button 1`] = `
<div>
  <div
    class="flex flex-col items-start"
  >
    <div>
      Fugiat deleniti quia doloremque. Et sed ut facilis et dolorem non. Qui neque velit quis dolore similique accusantium officiis adipisci. Dolorum cum et totam eveniet culpa qui quibusdam ad facere. Rerum distinctio atque consequatur amet iure sequi voluptas dolorem. Suscipit non quam. Deserunt omnis sint similique cum aliquid est sunt odio nemo. Inventore est ut ut magnam ea dolores ipsum ea recusandae. Et facere cumque quae fuga dolor ipsa dolor. Unde sit cupiditate. Sint aspernatur ullam a est repellendus ducimus. Maxime id est dolor nobis fuga. Voluptatem veniam maiores aut laborum perferendis harum quo minus. Consequatur natus nihil dicta et. Placeat et aut earum harum a. Impedit occaecati magni fuga et cumque excepturi enim voluptas. Quasi ipsum ut. Quia exercitationem sequi voluptates delectus ullam laborum. Porro et consequuntur harum. Hic praesentium maiores qui asperiores qui vero accusantium nemo voluptas. Sequi voluptatum perferendis aperiam qui assumenda velit necessitatibus. Adipisci animi molestias aspernatur quas vel consequatur veniam. Nisi voluptas eos. Facere laboriosam rerum quidem possimus nisi sit totam veniam. Repellendus magni et similique consequatur vel officiis ullam. Est quidem ullam eum voluptas porro et voluptates doloribus. Aut minima reprehenderit commodi nemo ducimus atque. Minima neque natus consequatur pariatur odit quaerat. Facere voluptatem quidem quia omnis et. Neque cum eveniet magnam vitae sint. Fugit quae consequatur quis quibusdam dolor. Quos odio quam nihil rerum quia assumenda sequi perspiciatis exercitationem. Reprehenderit quia molestiae aut quia voluptatem. Voluptatum dolore officiis facere ea hic nihil voluptatem ut. Consequuntur accusamus similique odit. Repellat culpa explicabo. In ex harum veniam. Corporis fuga saepe facilis. Ut labore similique molestias nihil. Aspernatur reprehenderit neque natus. Consequatur blanditiis odio ad blanditiis ab. Qui qui odit sit dignissimos voluptas veniam. Hic placeat corrupti natus qui eius magni autem iusto. Libero fugit consequatur voluptatibus nisi voluptas enim hic voluptate. Aliquam impedit consectetur. Praesentium libero quisquam. In eligendi labore. Numquam quis voluptas. Aut aut perferendis laborum autem. Sunt et dolore reprehenderit rerum dolorem itaque. In modi minima. Quisquam velit asperiores nemo ipsam iste amet at cumque rerum. Velit est id autem soluta deleniti voluptas.
    </div>
    <button
      class="btn btn-sm btn-link"
    >
      <svg
        class="w-5 h-5"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 320 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M177 159.7l136 136c9.4 9.4 9.4 24.6 0 33.9l-22.6 22.6c-9.4 9.4-24.6 9.4-33.9 0L160 255.9l-96.4 96.4c-9.4 9.4-24.6 9.4-33.9 0L7 329.7c-9.4-9.4-9.4-24.6 0-33.9l136-136c9.4-9.5 24.6-9.5 34-.1z"
        />
      </svg>
       Read less
    </button>
  </div>
</div>
`;

exports[`ReadMore snapshots a ReadMore with ReadMore button 1`] = `
<div>
  <div
    class="flex flex-col items-start"
  >
    <div>
      Libero nihil iusto deserunt qui id eveniet officia totam est. Quasi nihil itaque quos assumenda. Quia impedit rerum. Id rem at fugit alias. Quidem enim iste ab repellendus. Suscipit adipisci nulla odit debitis est aut. Non quia ratione itaque molestias. Libero aspernatur odio corporis. Repudiandae aut ipsam soluta minus vel vel. Autem modi quos qui eaque voluptas ea quibusdam totam. Officiis et aut est. Delectus voluptas est non fugit. Numquam dolor fugit ipsa consequatur et repellat.
    </div>
    <button
      class="btn btn-sm btn-link"
    >
      <svg
        class="w-5 h-5"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 320 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M143 352.3L7 216.3c-9.4-9.4-9.4-24.6 0-33.9l22.6-22.6c9.4-9.4 24.6-9.4 33.9 0l96.4 96.4 96.4-96.4c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9l-136 136c-9.2 9.4-24.4 9.4-33.8 0z"
        />
      </svg>
       Read more
    </button>
  </div>
</div>
`;

exports[`ReadMore snapshots a ReadMore without ReadMore button 1`] = `
<div>
  <div
    class="flex flex-col items-start"
  >
    <div>
      Libero nihil iusto deserunt qui id eveniet officia totam est. Quasi nihil itaque quos assumenda. Quia impedit rerum. Id rem at fugit alias. Quidem enim iste ab repellendus. Suscipit adipisci nulla odit debitis est aut. Non quia ratione itaque molestias. Libero aspernatur odio corporis. Repudiandae aut ipsam soluta minus vel vel. Autem modi quos qui eaque voluptas ea quibusdam totam. Officiis et aut est. Delectus voluptas est non fugit. Numquam dolor fugit ipsa consequatur et repellat.
    </div>
  </div>
</div>
`;

exports[`ReadMore with abnormal state snapshots a ReadMore when error 1`] = `
<div>
  <div
    class="alert alert-error shadow-lg"
  >
    <div>
      <svg
        class="w-6 h-6"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M8.6 1c1.6.1 3.1.9 4.2 2 1.3 1.4 2 3.1 2 5.1 0 1.6-.6 3.1-1.6 4.4-1 1.2-2.4 2.1-4 2.4-1.6.3-3.2.1-4.6-.7-1.4-.8-2.5-2-3.1-3.5C.9 9.2.8 7.5 1.3 6c.5-1.6 1.4-2.9 2.8-3.8C5.4 1.3 7 .9 8.6 1zm.5 12.9c1.3-.3 2.5-1 3.4-2.1.8-1.1 1.3-2.4 1.2-3.8 0-1.6-.6-3.2-1.7-4.3-1-1-2.2-1.6-3.6-1.7-1.3-.1-2.7.2-3.8 1-1.1.8-1.9 1.9-2.3 3.3-.4 1.3-.4 2.7.2 4 .6 1.3 1.5 2.3 2.7 3 1.2.7 2.6.9 3.9.6zM7.9 7.5L10.3 5l.7.7-2.4 2.5 2.4 2.5-.7.7-2.4-2.5-2.4 2.5-.7-.7 2.4-2.5-2.4-2.5.7-.7 2.4 2.5z"
          fill-rule="evenodd"
        />
      </svg>
      <span>
        Network Error!
      </span>
    </div>
  </div>
</div>
`;

exports[`ReadMore with abnormal state snapshots a ReadMore when loading 1`] = `
<div>
  <div
    class="flex flex-col items-start"
  >
    <div>
      Libero nihil iusto deserunt qui id eveniet officia totam est. Quasi nihil itaque quos assumenda. Quia impedit rerum. Id rem at fugit alias. Quidem enim iste ab repellendus. Suscipit adipisci nulla odit debitis est aut. Non quia ratione itaque molestias. Libero aspernatur odio corporis. Repudiandae aut ipsam soluta minus vel vel. Autem modi quos qui eaque voluptas ea quibusdam totam. Officiis et aut est. Delectus voluptas est non fugit. Numquam dolor fugit ipsa consequatur et repellat.
    </div>
    <button
      class="btn btn-sm btn-link loading"
    >
      Loading...
    </button>
  </div>
</div>
`;
